packages/@lwc/engine-dom/src/apis/build-custom-element-constructor.ts

```diff
@@ -45,20 +46,36 @@ export function deprecatedBuildCustomElementConstructor(
     return Ctor.CustomElementConstructor;
 }
 
+// Note: WeakSet is not supported in IE11, and the polyfill is not performant enough.
+//       This WeakSet usage is valid because this functionality is not meant to run in IE11.
+const hydratedCustomElements = new WeakSet<Element>();
+
 export function buildCustomElementConstructor(Ctor: ComponentConstructor): HTMLElementConstructor {
     const HtmlPrototype = getComponentHtmlPrototype(Ctor);
 
     return class extends HtmlPrototype {
         constructor() {
             super();
-            createVM(this, Ctor, {
-                mode: 'open',
-                owner: null,
-                tagName: this.tagName,
-            });
+
+            if (this.isConnected) {
+                // this if block is hit when there's already an un-upgraded element in the DOM with the same tag name.
+                hydrateComponent(this, Ctor, {});
+                hydratedCustomElements.add(this);
+            } else {
+                createVM(this, Ctor, {
+                    mode: 'open',
+                    owner: null,
+                    tagName: this.tagName,
+                });
+            }
         }
         connectedCallback() {
-            connectRootElement(this);
+            if (hydratedCustomElements.has(this)) {
```

Comment:
Is the only reason of this weakset to avoid calling connected twice?

Reply:
well, not just to avoid calling connectedCallback twice. [`connectRootElement`](https://github.com/salesforce/lwc/blob/master/packages/@lwc/engine-core/src/framework/vm.ts#L194-L209) does more than calling the `connectedCallback`.